src/main/java/org/jenkinsci/plugins/workflow/cps/nodes/StepDescriptorCache.java

```diff
@@ -55,26 +56,30 @@ public static void invalidateGlobalCache() {
     }
 
     public void invalidateAll() {
-        descriptorCache.clear();
+        store.clear();
     }
 
-    private final Memoizer<String, StepDescriptor> descriptorCache = new Memoizer<String, StepDescriptor>() {
-
-        public StepDescriptor compute(String descriptorId) {
-            Jenkins j = Jenkins.getInstance();
-            if (descriptorId != null && j != null) {
-                return (StepDescriptor) j.getDescriptor(descriptorId);
-            }
-            return null;
-        }
-    };
+    private final ConcurrentHashMap<String, StepDescriptor> store = new ConcurrentHashMap<String, StepDescriptor>();
 
     @CheckForNull
     public StepDescriptor getDescriptor(String descriptorId) {
-        if (descriptorId != null) {
-            return descriptorCache.get(descriptorId);
+        if (descriptorId == null) {
+            return null;
         }
-        return null;
 
+        StepDescriptor v = store.get(descriptorId);
+        if(v!=null) return v;
+
+        synchronized (this) {
```

Comment:
@jglick Look closely - this isn't reliant on field access, it does not suffer from double-checked locking, and is adapted from Memoizer itself.